mountpoint-s3/src/fs.rs

```diff
@@ -585,6 +589,37 @@ where
         }
     }
 
+    pub async fn fsync(&self, _ino: InodeNo, fh: u64, _datasync: bool) -> Result<(), libc::c_int> {
+        let file_handle = {
+            let file_handles = self.file_handles.read().await;
+            match file_handles.get(&fh) {
+                Some(handle) => handle.clone(),
+                None => return Err(libc::EBADF),
+            }
+        };
+        let mut request = match &file_handle.typ {
+            FileHandleType::Write(request) => request.lock().await,
+            FileHandleType::Read { .. } => return Ok(()),
+        };
+        let Some(in_progress_request) = request.take() else {
+            error!(key=?file_handle.full_key, "upload already completed or aborted");
```

Comment:
I think we talked about tracking a little more state in the handle so we could distinguish these two cases. Is that worth doing?

Reply:
Also, should be `warn!`

Reply:
I've added `UploadState` to keep track of failure vs completion.